// rethinkdb/cursor.js
// Generated by CoffeeScript 1.7.0
var ArrayResult, Cursor, IterableResult, ar, aropt, deconstructDatum, err, mkErr, nextCbCheck, pb, setImmediate, util, varar,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

err = require('./errors');

util = require('./util');

pb = require('./protobuf');

ar = util.ar;

varar = util.varar;

aropt = util.aropt;

deconstructDatum = util.deconstructDatum;

mkErr = util.mkErr;

if (typeof setImmediate === "undefined" || setImmediate === null) {
  setImmediate = function(cb) {
    return setTimeout(cb, 0);
  };
}

IterableResult = (function() {
  function IterableResult() {}

  IterableResult.prototype.hasNext = function() {
    throw "Abstract Method";
  };

  IterableResult.prototype.next = function() {
    throw "Abstract Method";
  };

  IterableResult.prototype.each = varar(1, 2, function(cb, onFinished) {
    var brk, n;
    if (typeof cb !== 'function') {
      throw new err.RqlDriverError("First argument to each must be a function.");
    }
    if ((onFinished != null) && typeof onFinished !== 'function') {
      throw new err.RqlDriverError("Optional second argument to each must be a function.");
    }
    brk = false;
    n = (function(_this) {
      return function() {
        if (!brk && _this.hasNext()) {
          return _this.next(function(err, row) {
            brk = cb(err, row) === false;
            return n();
          });
        } else if (onFinished != null) {
          return onFinished();
        }
      };
    })(this);
    return n();
  });

  IterableResult.prototype.toArray = ar(function(cb) {
    var arr;
    if (typeof cb !== 'function') {
      throw new err.RqlDriverError("Argument to toArray must be a function.");
    }
    arr = [];
    if (!this.hasNext()) {
      cb(null, arr);
    }
    return this.each((function(_this) {
      return function(err, row) {
        if (err != null) {
          cb(err);
        } else {
          arr.push(row);
        }
        if (!_this.hasNext()) {
          return cb(null, arr);
        }
      };
    })(this));
  });

  return IterableResult;

})();

Cursor = (function(_super) {
  __extends(Cursor, _super);

  Cursor.prototype.stackSize = 100;

  function Cursor(conn, token, opts, root) {
    this._conn = conn;
    this._token = token;
    this._opts = opts;
    this._root = root;
    this._responses = [];
    this._responseIndex = 0;
    this._outstandingRequests = 1;
    this._iterations = 0;
    this._endFlag = false;
    this._contFlag = false;
    this._cont = null;
    this._cbQueue = [];
  }

  Cursor.prototype._addResponse = function(response) {
    this._responses.push(response);
    this._outstandingRequests -= 1;
    pb.ResponseTypeSwitch(response, {
      "SUCCESS_PARTIAL": (function(_this) {
        return function() {
          return _this._endFlag = false;
        };
      })(this)
    }, (function(_this) {
      return function() {
        return _this._endFlag = true;
      };
    })(this));
    this._contFlag = false;
    this._promptNext();
    return this;
  };

  Cursor.prototype._getCallback = function() {
    var cb, immediateCb;
    this._iterations += 1;
    cb = this._cbQueue.shift();
    if (this._iterations % this.stackSize === this.stackSize - 1) {
      immediateCb = (function(err, row) {
        return setImmediate(function() {
          return cb(err, row);
        });
      });
      return immediateCb;
    } else {
      return cb;
    }
  };

  Cursor.prototype._handleRow = function() {
    var cb, response, row;
    response = this._responses[0];
    row = deconstructDatum(response.response[this._responseIndex], this._opts);
    cb = this._getCallback();
    this._responseIndex += 1;
    if (this._responseIndex === response.response.length) {
      this._responses.shift();
      this._responseIndex = 0;
    }
    return cb(null, row);
  };

  Cursor.prototype._promptNext = function() {
    var cb, response;
    while (this._cbQueue[0] != null) {
      if (!this.hasNext()) {
        cb = this._getCallback();
        cb(new err.RqlDriverError("No more rows in the cursor."));
      } else {
        response = this._responses[0];
        if (this._responses.length === 1) {
          this._promptCont();
          if (!this._endFlag && (response.response != null) && this._responseIndex === response.response.length - 1) {
            return;
          }
        }
        pb.ResponseTypeSwitch(response, {
          "SUCCESS_PARTIAL": (function(_this) {
            return function() {
              return _this._handleRow();
            };
          })(this),
          "SUCCESS_SEQUENCE": (function(_this) {
            return function() {
              if (response.response.length === 0) {
                return _this._responses.shift();
              } else {
                return _this._handleRow();
              }
            };
          })(this),
          "COMPILE_ERROR": (function(_this) {
            return function() {
              _this._responses.shift();
              cb = _this._getCallback();
              return cb(mkErr(err.RqlCompileError, response, _this._root));
            };
          })(this),
          "CLIENT_ERROR": (function(_this) {
            return function() {
              _this._responses.shift();
              cb = _this._getCallback();
              return cb(mkErr(err.RqlClientError, response, _this._root));
            };
          })(this),
          "RUNTIME_ERROR": (function(_this) {
            return function() {
              _this._responses.shift();
              cb = _this._getCallback();
              return cb(mkErr(err.RqlRuntimeError, response, _this._root));
            };
          })(this)
        }, (function(_this) {
          return function() {
            _this._responses.shift();
            cb = _this._getCallback();
            return cb(new err.RqlDriverError("Unknown response type for cursor"));
          };
        })(this));
      }
    }
  };

  Cursor.prototype._promptCont = function() {
    if (!this._contFlag && !this._endFlag) {
      this._contFlag = true;
      this._outstandingRequests += 1;
      return this._conn._continueQuery(this._token);
    }
  };

  Cursor.prototype.hasNext = ar(function() {
    return (this._responses[0] != null) && this._responses[0].response.length > 0;
  });

  Cursor.prototype.next = ar(function(cb) {
    nextCbCheck(cb);
    this._cbQueue.push(cb);
    return this._promptNext();
  });

  Cursor.prototype.close = ar(function() {
    if (!this._endFlag) {
      this._outstandingRequests += 1;
      return this._conn._endQuery(this._token);
    }
  });

  Cursor.prototype.toString = ar(function() {
    return "[object Cursor]";
  });

  return Cursor;

})(IterableResult);

ArrayResult = (function(_super) {
  __extends(ArrayResult, _super);

  function ArrayResult() {
    return ArrayResult.__super__.constructor.apply(this, arguments);
  }

  ArrayResult.prototype.stackSize = 100;

  ArrayResult.prototype.hasNext = ar(function() {
    if (this.__index == null) {
      this.__index = 0;
    }
    return this.__index < this.length;
  });

  ArrayResult.prototype.next = ar(function(cb) {
    var self;
    nextCbCheck(cb);
    if (this.__index == null) {
      this.__index = 0;
    }
    if (this.hasNext() === true) {
      self = this;
      if (self.__index % this.stackSize === this.stackSize - 1) {
        return setImmediate(function() {
          return cb(null, self[self.__index++]);
        });
      } else {
        return cb(null, self[self.__index++]);
      }
    } else {
      return cb(new err.RqlDriverError("No more rows in the cursor."));
    }
  });

  ArrayResult.prototype.toArray = ar(function(cb) {
    if (this.__index != null) {
      return cb(null, this.slice(this.__index, this.length));
    } else {
      return cb(null, this);
    }
  });

  ArrayResult.prototype.close = function() {
    return this;
  };

  ArrayResult.prototype.makeIterable = function(response) {
    var method, name, _ref;
    response.__proto__ = {};
    _ref = ArrayResult.prototype;
    for (name in _ref) {
      method = _ref[name];
      if (name !== 'constructor') {
        response.__proto__[name] = method;
      }
    }
    response.__proto__.__proto__ = [].__proto__;
    return response;
  };

  return ArrayResult;

})(IterableResult);

nextCbCheck = function(cb) {
  if (typeof cb !== 'function') {
    throw new err.RqlDriverError("Argument to next must be a function.");
  }
};

module.exports.deconstructDatum = deconstructDatum;

module.exports.Cursor = Cursor;

module.exports.makeIterable = ArrayResult.prototype.makeIterable;
